### PR TITLE
Adewyer working

### PR DESCRIPTION
**Pull request by adewyer**
UQ CODE

### CODE REVIEW

File: kinbot/hindered_rotors.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import time
 import logging
 import numpy as np
-import matplotlib.pyplot as plt
+#import matplotlib.pyplot as plt
```
Comment (adewyer) [inline kinbot/hindered_rotors.py:4]: import was causing error because of incompatibilities between imports.

File: kinbot/postprocess.py [hunk 2]
```diff
@@ -6,16 +6,132 @@
...
+
+
+def create_sql_db(species):
+    db = str(species.chemid) + '_sql.db'
+    if path.exists(db):
+        print("Path exists: {}".format(db))
+    else:
+        conn = sq.create_connection(db)
+        print("creating kinbot table")
+        sq.create_kinbot_table(conn)
```
Comment (adewyer) [inline kinbot/postprocess.py:37]: delete sq_db inside of create

File: kinbot/sq.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                 ); '''
+
+    try:
+        dbc = conn.cursor()
+        dbc.execute(sql)
+    except Error as err:
+        print(err)
+
+
+def create_kinbot(conn, kinbot_table):
```
Comment (adewyer) [inline kinbot/sq.py:106]: rename create_entry

File: kinbot/hindered_rotors.py (hunk 1 above)
Comment (adewyer) [inline kinbot/hindered_rotors.py:4]: matplot was causing import errors w/ another package

File: kinbot/homolytic_scissions.py [hunk 4]
```diff
@@ -71,14 +71,25 @@ def find_homolytic_scissions(self):
                         # get the elements of the current bond under consideration, and put them
                         # in alphabetical order
                         bond_elements = sorted([self.species.atom[i], self.species.atom[j]])
+                        b_e = []
```
Comment (adewyer) [inline kinbot/homolytic_scissions.py:74]: Code updated to make sure that user input homolytic scission elements are sorted and listed in the same was as the `bond_elements` variable. 

For example
user input = `OC` --> needs to be sorted to be listed as `CO` in the same way that  bond_elements = `CO`. This check helps ensure that the user input matches the `bond_elements` variable when kinbot checks. 

Without this code, a user input of `OC` will be read as `OC != CO` even though `CO` and `OC` should be considered as the same type of bond being targeted for homolytic scission reactions.

File: kinbot/zmatrix.py [hunk 5]
```diff
@@ -1,4 +1,3 @@
-from __future__ import print_function
```
Comment (adewyer) [inline kinbot/zmatrix.py:1]: clean up from flake8 - import unused

File: kinbot/reaction_finder.py [hunk 6]
```diff
@@ -127,7 +127,7 @@ def find_reactions(self):
                           'r13_insertion_RSR': self.search_r13_insertion_RSR, 
                           'beta_delta': self.search_beta_delta, 
                           'h2_elim': self.search_h2_elim,
-                          'hom_sci': self.search_hom_sci,
+                          # 'hom_sci': self.search_hom_sci,
```
Comment (adewyer) [inline kinbot/reaction_finder.py:130]: currently missing this function in `reactions/` directory

File: kinbot/optimize.py [hunk 7]
```diff
@@ -222,11 +222,11 @@ def do_optimization(self):
                                 err, fr = self.qc.get_qc_freq(self.job_high, self.species.natom)
                                 if self.species.natom == 1:
                                     freq_ok = 1
-                                elif len(fr) == 1 and fr[0] == 0:
+                                elif len(fr) == 1 and float(fr[0]) == 0:
```
Comment (adewyer) [inline kinbot/optimize.py:225]: freq were being returned as strings. So swapped to float before comparing to 0.

File: tpl/pbs_mess_uq.tpl [hunk 8]
```diff
@@ -1,2 +0,0 @@
-cd ${{PBS_O_WORKDIR}}/me
```
Comment (adewyer) [inline tpl/pbs_mess_uq.tpl:1]: file deleted because all mess files will have a numbering regardless of whether or not UQ is turned on to simpliify things.

File: kinbot/uncertaintyAnalysis.py [hunk 9]
```diff
@@ -1,106 +0,0 @@
-from __future__ import division
```
Comment (adewyer) [inline kinbot/uncertaintyAnalysis.py:1]: file was renamed to `uncertainty_analysis.py`

File: kinbot/kb.py [hunk 10]
```diff
@@ -127,14 +139,12 @@ def main():
...
         logging.error('Found imaginary frequency for initial structure.')
-        return
 
     # characterize again and look for differences
     well0.characterize(dimer=par['dimer'])
     well0.name = str(well0.chemid)
     if well0.name != start_name:
-        logging.error('The first well optimized to a structure different from the input.')
-        return
+       logging.error('The first well optimized to a structure different from the input.')
```
Comment (adewyer) [inline kinbot/kb.py:147]: This turn off is temporary. Will change code back

File: kinbot/kb.py [hunk 11]
```diff
@@ -116,7 +123,12 @@ def main():
     #            time.sleep(1)
 
     # start the initial optimization of the reactant
-    logging.info('Starting optimization of intial well')
+    logging.info('Starting optimization of intial well {}'.format(well0.chemid))
+    # for debugging
+    tmp=open("geom0.xyz",'w')
+    tmp.write("chemid: {}\natom: {}\nbond: {}\n{}".format(well0.chemid, well0.atom, well0.bond, str(well0.geom)))
+    tmp.close()
```
Comment (adewyer) [inline kinbot/kb.py:130]: to remove

File: kinbot/kb.py [hunk 12]
```diff
@@ -100,6 +100,13 @@ def main():
     well0.characterize(dimer=par['dimer'])
     well0.name = str(well0.chemid)
     start_name = well0.name
+    # for debugging
+    f=open("chemid", 'w')
+    f.write(str(well0.chemid))
+    f.close()
```
Comment (adewyer) [inline kinbot/kb.py:106]: to remove

File: kinbot/kb.py [hunk 13]
```diff
@@ -192,14 +202,23 @@ def main():
...
+        if well0.reac_ts_done[index] == -1:
+            ts = well0.reac_obj[index].ts
+            postprocess.create_sql_db_entry(well0, well0, ts, qc, par, 2, index)
+            for prod in well0.reac_obj[index].products:
+                postprocess.create_sql_db_entry(well0, prod, 'None', qc, par, 1, index)
     postprocess.createSummaryFile(well0, qc, par)
     postprocess.createPESViewerInput(well0, qc, par)
     postprocess.creatMLInput(well0, qc, par)
+    uq_obj = UQ(par)
+    # uq_obj.format_uqtk_data()  # needs editing
```
Comment (adewyer) [inline kinbot/kb.py:221]: NEEDS WORK

File: kinbot/reaction_finder.py (hunk 6 above)
Comment (adewyer) [reply kinbot/reaction_finder.py:129]: fixed

File: kinbot/reaction_generator.py [hunk 14]
```diff
@@ -622,7 +625,8 @@ def test_stereochem_pybel(self, pybelErr, obj):
         """
 
         if pybelErr == 1:
-            return 0
```
Comment (adewyer) [inline kinbot/reaction_generator.py:625]: no variable to return.

File: kinbot/reaction_generator.py [hunk 15]
```diff
@@ -634,7 +638,7 @@ def test_stereochem_pybel(self, pybelErr, obj):
                                     .format(prod.chemid, obj.instance_name))
             else:
                 continue
-        return 0
```
Comment (adewyer) [inline kinbot/reaction_generator.py:637]: no variable to return

File: kinbot/mess.py [hunk 16]
```diff
@@ -316,14 +353,14 @@ def write_input(self, qc):
             for rxn in barrierless_blocks:
                 barrierless += barrierless_blocks[rxn] + divider
 
-            dum = self.dummytpl.format(barrier='tsd', reactant=self.well_names[self.species.chemid], dummy='d1')
+            # dum = self.dummytpl.format(barrier='tsd', reactant=self.well_names[self.species.chemid], dummy='d1')
```
Comment (adewyer) [inline kinbot/mess.py:356]: not sure  why we need this line?

File: kinbot/mess.py [hunk 17]
```diff
@@ -316,14 +353,14 @@ def write_input(self, qc):
...
 
             mess_iter = "{0:04d}".format(uq_iter)
 
             with open('me/mess_%s.inp' % mess_iter, 'w') as f_out:
                 f_out.write(header + divider + wells + bimols + tss + termols + barrierless + divider + 'End ! end kinetics\n')
 
-        uq_obj.format_uqtk_data() 
+        # uq_obj.format_uqtk_data()  # needs editing
```
Comment (adewyer) [inline kinbot/mess.py:363]: Formatting the data needs to be done to make the UQTk readable file.

File: kinbot/mess.py [hunk 18]
```diff
@@ -615,13 +664,17 @@ def write_barrier(self, reaction, index, left_zeroenergy, right_zeroenergy, barr
                                                   long_rxn_name=long_rxn_name,
                                                   model=variational)
         else:
+            if self.par['pes']:
+                nfreq = '{nfreq}'
```
Comment (adewyer) [inline kinbot/mess.py:668]: I think this can be set not paramaterized

File: kinbot/mess.py [hunk 19]
```diff
@@ -653,21 +706,24 @@ def run(self):
         while uq_iter < self.par['uq_n']:
             self.write_submitscript(submitscript, uq_iter)
             while len(pids) > self.par['uq_max_runs']:
-                time.sleep(5)
+                time.sleep(1)
                 for pid in pids:
                     stat = self.check_running(pid)
-                    if stat == 0:
```
Comment (adewyer) [inline kinbot/mess.py:659]: numbering swapped from rosalind to menten at one point.
Double check

File: kinbot/pes.py [hunk 20]
```diff
@@ -912,17 +917,33 @@ def create_mess_input(par, wells, products, reactions, barrierless,
...
+
+    for b in barrierless:
+        pathway.append(b[0])
+        for prod in b[2]:
+            pathway.append(prod)
+        pathway.append(rxn[2])
+    try:
+        all_structures[rxn[1]] = pathway
+    except:
+        pass
```
Comment (adewyer) [inline kinbot/pes.py:940]: This section will be deleted after testing

File: kinbot/conformers.py [hunk 21]
```diff
@@ -441,14 +441,13 @@ def check_conformers(self, wait=0):
                                     err = -1
                             if err == 0:
                                 lowest_totenergy = energy + zpe
-                                if self.species.wellorts:
-                                    base_imag_freq = freq[0]
                         if energy + zpe < lowest_totenergy:
```
Comment (adewyer) [inline kinbot/conformers.py:444]: placement of `base_imag_freq` was causing an error about the variable being called before being assigned below. 

So the line is moved to below so that it can be assigned before being called.

File: kinbot/kb.py [hunk 22]
```diff
@@ -126,15 +130,13 @@ def main():
         logging.error('Error with initial structure optimization.')
         return
     if any(well0.freq[i] <= 0 for i in range(len(well0.freq))):
-        logging.error('Found imaginary frequency for initial structure.')
-        return
+        logging.error('Found imaginary frequency (freq # {}, value = {}) for initial structure.'.format(i, well0.freq[i]))
```
Comment (adewyer) [inline kinbot/kb.py:134]: Occassionaly a structure has a single small imaginary freq. related to a Me rotation or some other small movement that the optimizer cannot eliminate. So instead of completely killing the entire calculation it continues running but the error is noted in the log file. If the frequency is detrimental to the calculation then the user can check, but  the only time I have seen this is when the error comes from a small imaginary freq that is related to a Me group or something similar.

File: kinbot/kb.py [hunk 23]
```diff
@@ -126,15 +130,13 @@ def main():
...
 
     # characterize again and look for differences
     well0.characterize(dimer=par['dimer'])
     well0.name = str(well0.chemid)
     if well0.name != start_name:
-        logging.error('The first well optimized to a structure different from the input.')
-        return
+       logging.error('The first well optimized to a structure different from the input.')
```
Comment (adewyer) [inline kinbot/kb.py:139]: Again, i let the calculation continue here, but noting the error in the log file.

File: tpl/mess_hinderedrotor.tpl [hunk 24]
```diff
@@ -1,6 +1,4 @@
       Rotor     Hindered
-      Geometry[angstrom]          {natom}
-{geom}
```
Comment (adewyer) [inline tpl/mess_hinderedrotor.tpl:3]: This line caused errors in MESS, you can only have one instance of a `Geometry` section per block.

File: kinbot/uncertainty_analysis.py [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+    def write_uqtk_data(self, propertyType, normfactor, species, uq_iter):
+        file = "uq_" + str(species) + "_" + str(propertyType) + ".txt"
+        with open(file, 'a') as fi:
+            fi.write("{} | {}\n".format(uq_iter, normfactor))
+            if uq_iter == self.par["uq_n"] + 1:
+                fi.write("\n")
+
+        return 0
+
+    def format_uqtk_data(self):
```
Comment (adewyer) [inline kinbot/uncertainty_analysis.py:124]: `format_uqtk_data` is not working yet.

File: kinbot/uncertainty_analysis.py [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+            factor = np.exp(random.uniform(np.log(1./self.hirUQ), np.log(self.hirUQ)))
+            normfactor = np.log(factor) / np.log(self.hirUQ)
+            if normfactor < -1 or normfactor > 1:
+                e = "ERROR"
+            else:
+                e = ''
+
+        return factor, normfactor
+
+    def calc_factor(self, propertyType, species, uq_iter, runUQ):
```
Comment (adewyer) [inline kinbot/uncertainty_analysis.py:60]: the `runUQ` parameter is meant to establish if the UQ values should be set yet. For instance, in a PES run, we need to wait to calculate the values until we get past all of the KinBot steps and piece together all of the mess files from each kinbot run.

In the case of a KinBot run, the mess files can be generated when the KinBot run is done because we don't need to account for piecing a MESS file together from multiple kinbot runs at once. 

Basically if `runUQ=0` then place holders for energies are set during kinbot runs, and are filled in later once all the mess files (with place holders) have been created.

If `runUQ=1` then the mess files are created wth the UQ values input into them immediately without any place holders.

File: tpl/barrierless_prod_mess.tpl [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+            0    2
+
+        ZeroEnergy[kcal/mol]        0.0
+    End ! end fragment
+
+        GroundEnergy[kcal/mol]      {prod_energy}
+
+    
+    End ! end bimolecular
+!****************************************
```
Comment (adewyer) [inline tpl/barrierless_prod_mess.tpl:45]: The user needs to fill in this file with everything except `{prod_energy}` and have it in their kinbot or  pes main directory.

Comment (adewyer) [reply tpl/barrierless_prod_mess.tpl:45]: The template file can be found in `/tpl` but the user needs to copy the file into their main kinbot or pes run directory and edit everything in the file except for the `{prod_energy}` value.

File: tpl/barrierless_mess.tpl [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+        {freqs fragment 2}
+
+    ! no tunneling bc no barrier
+
+    ZeroEnergy[kcal/mol]                {prod_energy}
+    ! ElectronicEnergy[kcal/mol]          0
+    ElectronicLevels[1/cm]              1
+          0     2
+End ! end barrierless reaction
+!****************************************
```
Comment (adewyer) [inline tpl/barrierless_mess.tpl:20]: The template file can be found in /tpl but the user needs to copy the file into their main kinbot or pes run directory and edit everything in the file except for the {prod_energy} and {mc_flux.out} values.

File: kinbot/parameters.py [hunk 29]
```diff
@@ -258,26 +258,47 @@ def __init__(self, inpfile=None):
             # MESMER specific keywords
             'mesmer_command': 'mesmer',
 
+            # Barrierless MESS keywords
+            # 0 = no barrierless rxn, 1 == include barrierless rxn templates
+            'barrierless_rxn': 0,
```
Comment (adewyer) [inline kinbot/parameters.py:263]: Whether or not to include barrierless templates in the output mess files.

File: kinbot/mess.py [hunk 30]
```diff
@@ -196,6 +239,7 @@ def write_input(self, qc):
 
         # write the mess input for the different blocks
         for uq_iter in range(self.par['uq_n']):
+            header = self.write_header(uq_iter)
```
Comment (adewyer) [inline kinbot/mess.py:242]: moved create header into uq portion of code so that epsilon/sigma/energy relaxation factors can be edited for each iteration.

File: kinbot/mess.py [hunk 31]
```diff
@@ -89,18 +119,32 @@ def write_header(self):
...
+        energy = barrierless_energy
+        if uq_iter > 0:
+            barrierless_energy_factor, barrierless_energy_normfactor = uq_obj.calc_factor('energy', 'barrierless', uq_iter, 1)
+            energy = energy + barrierless_energy_factor
+            uq_obj.write_uqtk_data("barrierless_energy", barrierless_energy_normfactor, 'barrierless', uq_iter)
+            print(energy, barrierless_energy_factor)
+        print(energy)
+        barrierless_block = self.barrierless_template.format(energy=energy, flux_file=self.par['barrierless_states_file'])
+        barrierless_prod_block = self.barrierless_prod_template.format(energy=energy)
+        return barrierless_block, barrierless_prod_block
```
Comment (adewyer) [inline kinbot/mess.py:147]: lines 134-147 create the barrierless blocks for the ROO* --> O2 + R* reactions.

File: kinbot/mess.py [hunk 32]
```diff
@@ -315,15 +372,20 @@ def write_input(self, qc):
             barrierless = ''
             for rxn in barrierless_blocks:
                 barrierless += barrierless_blocks[rxn] + divider
+            if self.par['barrierless_rxn'] == 1:
+                barrierless_block, barrierless_prod_block = self.create_barrierless_block(self.par['barrierless_energy'], uq_obj, uq_iter)
 
-            dum = self.dummytpl.format(barrier='tsd', reactant=self.well_names[self.species.chemid], dummy='d1')
```
Comment (adewyer) [inline kinbot/mess.py:319]: deleted,  not sure this line was necessary.

File: kinbot/mess.py [hunk 33]
```diff
@@ -367,24 +429,27 @@ def write_bimol(self, prod_list, well_add, freqFactor, uq_iter, bless=0):
...
         smi = []
-        for nsp, species in enumerate(prod_list):
+        for i, species in enumerate(prod_list):
             smi.append(species.smiles)
             if species.natom > 1:
-
                 if self.par['pes']:
                     name = '{{fr_name_{}}}'.format(species.chemid)
+                    freq = '{{freq_{}}}'.format(i) 
+                    nfreq = '{{nfreq_{}}}'.format(i)
```
Comment (adewyer) [inline kinbot/mess.py:438]: When PES is turned on the `freq` and `nfreq` values become keywords to fill in during the final PES MESS file creation step.

File: kinbot/mess.py [hunk 34]
```diff
@@ -537,7 +602,13 @@ def write_barrier(self, reaction, index, left_zeroenergy, right_zeroenergy, barr
...
             tun = ''
         else:
-            tun = self.tunneltpl.format(cutoff=min(left_zeroenergy, right_zeroenergy),
+            if self.par['pes']:
+                tun = self.tunneltpl.format(cutoff='{cutoff}',
+                                        imfreq='{imfreq}',
+                                        welldepth1='{left_zero}',
+                                        welldepth2='{right_zero}')
+            else:
+                tun = self.tunneltpl.format(cutoff=min(left_zeroenergy, right_zeroenergy),
```
Comment (adewyer) [inline kinbot/mess.py:611]: tunneling left as keywords to fill in for PES, so that for each uq iteration new energy levels can be calculated and input as needed.

File: kinbot/pes.py [hunk 35]
```diff
@@ -931,84 +1010,187 @@ def create_mess_input(par, wells, products, reactions, barrierless,
...
-                        EnergyRelaxationFactor=par['EnergyRelaxationFactor'],
-                        EnergyRelaxationPower=par['EnergyRelaxationPower'],
-                        EnergyRelaxationExponentCutoff=par['EnergyRelaxationExponentCutoff'],
-                        e_coll=constants.epsilon[par['collider']],
-                        s_coll=constants.sigma[par['collider']],
-                        m_coll=constants.mass[par['collider']],
-                        e_well=par['epsilon'],
-                        s_well=par['sigma'],
-                        m_well=well0.mass,
-                        )
```
Comment (adewyer) [inline kinbot/pes.py:958]: Moved this to its own function.

File: kinbot/pes.py [hunk 36]
```diff
@@ -931,84 +1010,187 @@ def create_mess_input(par, wells, products, reactions, barrierless,
...
+                        freq_1 = freq_tmp
+                elif i == 1:
+                    if natoms == 1: 
+                        nfreq_tmp = nfreq_1
+                        nfreq_1 = nfreq_0
+                        nfreq_0 = nfreq_tmp
+                        freq_tmp = freq_1
+                        freq_1 = freq_0
+                        freq_0 = freq_tmp
+                f.close()
```
Comment (adewyer) [inline kinbot/pes.py:1166]: lines 1150-1166 ensure that the fragments are printed properly into the MESS file in the case where a single atom is part of the bimolecular prod.

File: kinbot/mess.py [hunk 37]
```diff
@@ -653,21 +728,24 @@ def run(self):
...
             pid = self.submit(submitscript)
             pids.append(pid)
 
             if self.par['uq_n'] < self.par['uq_max_runs']:
-                stat = 1
-                while stat != 0:
+                # stat = 1  # menten
+                stat = 0  #rosalind
+                # while stat != 0:  # menten
+                while stat != 1:  # rosalind
```
Comment (adewyer) [inline kinbot/mess.py:745]: Double check these values on menten/blodgett. At one point you needed to swap the value when using slurm vs pbs.

File: kinbot/uncertainty_analysis.py [hunk 38]
```diff
@@ -0,0 +1,179 @@
...
+            if stop_parsing == 0:
+                if char != " ":
+                    name_array.append(char)
+                elif char == " ":
+                    stop_parsing = 1
+            else:
+                pass
+        name_array.reverse()
+        name_string = ''.join(name_array)
+
```
Comment (adewyer) [inline kinbot/uncertainty_analysis.py:44]: delete lines 28-44?
